prow/config/agent.go

```diff
@@ -40,70 +47,249 @@ type Agent struct {
+				return fmt.Errorf("Failed to check is path %s is configmap mounted: %v", path, err)
+			} else if isCM {
+				cms.Insert(path)
+				// configmaps can't have nested directories
+				return filepath.SkipDir
+			} else {
+				dirs.Insert(path)
+				return nil
+			}
+		}
+		return nil
+	})
+	return cms, dirs, err
+}
+
+func watchConfigs(ca *Agent, prowConfig, jobConfig string, additionals ...func(*Config) error) error {
+	cmEventFunc := func() error {
+		c, err := Load(prowConfig, jobConfig, additionals...)
+		if err != nil {
+			return err
+		}
+		ca.Set(c)
+		return nil
+	}
+	// We may need to add more directories to be watched
+	dirsEventFunc := func(w *fsnotify.Watcher) error {
+		c, err := Load(prowConfig, jobConfig, additionals...)
+		if err != nil {
+			return err
+		}
+		ca.Set(c)
+		// TODO(AlexNPavel): Is there a chance that a ConfigMap mounted directory may appear without making a new pod? If yes, handle that.
+		_, dirs, err := ListCMsAndDirs(jobConfig)
+		if err != nil {
+			return err
+		}
+		for dir := range dirs {
+			// Adding a file or directory that already exists in fsnotify is a no-op, so it is safe to always run Add
+			if err := w.Add(dir); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	errFunc := func(err error, msg string) {
+		logrus.WithField("prowConfig", prowConfig).
+			WithField("jobConfig", jobConfig).
+			WithError(err).Error(msg)
+	}
+	cms := sets.NewString()
+	dirs := sets.NewString()
+	// TODO(AlexNPavel): allow empty jobConfig till fully migrate config to subdirs
 	if jobConfig != "" {
-		jobConfigStat, err := os.Stat(jobConfig)
+		stat, err := os.Stat(jobConfig)
 		if err != nil {
-			logrus.WithField("jobConfig", jobConfig).WithError(err).Error("Error loading job configs.")
-			return time.Time{}, err
+			return err
 		}
-
-		if jobConfigStat.ModTime().After(recentModTime) {
-			recentModTime = jobConfigStat.ModTime()
+		// TODO(AlexNPavel): allow single file jobConfig till fully migrate config to subdirs
+		if stat.IsDir() {
+			var err error
+			// jobConfig points to directories of configs that may be nested
+			cms, dirs, err = ListCMsAndDirs(jobConfig)
+			if err != nil {
+				return err
+			}
+		} else {
+			// If jobConfig is a single file, we handle it identically to how prowConfig is handled
+			if jobIsCMMounted, err := IsConfigMapMount(filepath.Dir(jobConfig)); err != nil {
+				return err
+			} else if jobIsCMMounted {
+				cms.Insert(filepath.Dir(jobConfig))
+			} else {
+				dirs.Insert(jobConfig)
+			}
 		}
 	}
-	return recentModTime, nil
+	// The prow config is always a single file
+	if prowIsCMMounted, err := IsConfigMapMount(filepath.Dir(prowConfig)); err != nil {
+		return err
+	} else if prowIsCMMounted {
+		cms.Insert(filepath.Dir(prowConfig))
+	} else {
+		dirs.Insert(prowConfig)
+	}
+	for cm := range cms {
+		runFunc, err := GetCMMountWatcher(cmEventFunc, errFunc, cm)
+		if err != nil {
+			return err
+		}
+		interrupts.Run(runFunc)
```

Comment:
Hm. This ways we start spawning watchers, then possibly hit an error and immediately return with that error, leaving some watchers running and some not. If that is acceptable and desired, we should accumulate errors and try to spawn the rest of the watchers. If not we should IMO prepare first and start spawning only if we encountered no errors.

Reply:
Fixed